app/scripts/lib/multichain-api/wallet-revokePermissions.js

```diff
@@ -0,0 +1,108 @@
+import { invalidParams, MethodNames } from '@metamask/permission-controller';
+import { isNonEmptyArray } from '@metamask/utils';
+import { RestrictedMethods } from '../../../../shared/constants/permissions';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from './caip25permissions';
+import { KnownCaipNamespace, parseScopeString } from './scope';
+
+export const revokePermissionsHandler = {
+  methodNames: [MethodNames.revokePermissions],
+  implementation: revokePermissionsImplementation,
+  hookNames: {
+    revokePermissionsForOrigin: true,
+    getPermissionsForOrigin: true,
+    updateCaveat: true,
+  },
+};
+
+/**
+ * Revoke Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.revokePermissionsForOrigin - A hook that revokes given permission keys for an origin
+ * @param options.getPermissionsForOrigin
+ * @param options.updateCaveat
+ * @returns A promise that resolves to nothing
+ */
+function revokePermissionsImplementation(
+  req,
+  res,
+  _next,
+  end,
+  { revokePermissionsForOrigin, getPermissionsForOrigin, updateCaveat },
+) {
+  const { params, origin } = req;
+
+  const param = params?.[0];
+
+  if (!param) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  // For now, this API revokes the entire permission key
+  // even if caveats are specified.
+  const permissionKeys = Object.keys(param).filter(
+    (name) => name !== Caip25EndowmentPermissionName,
+  );
+
+  if (!isNonEmptyArray(permissionKeys)) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  revokePermissionsForOrigin(permissionKeys);
+
+  const permissions = getPermissionsForOrigin(origin) || {};
+  const caip25endowment = permissions?.[Caip25EndowmentPermissionName];
+  const caip25caveat = caip25endowment?.caveats.find(
+    ({ type }) => type === Caip25CaveatType,
+  );
+
+  if (
+    process.env.BARAD_DUR &&
+    permissionKeys.includes(RestrictedMethods.eth_accounts) &&
+    caip25caveat
+  ) {
+    // should we remove accounts from required scopes? if so doesn't that mean we should
+    // just revoke the caip25endowment entirely?
```

Comment:
probably. I mean currently this revoke call isn't even incremental anyways

Reply:
Oh I see theres this weird case where CAIP-25 permissions can be valid without accounts. 

Sooo don't kill me since this is another case where I think I had woefully underspecified but I think we probably don't want to allow permissions without accounts associated. At least not for now? Let me ask what others think and come back on this.

Reply:
I suppose its possible we want this for Snaps 🤔 But I would think not for dapps. In any case this would still be a weird way to get to that state...

Reply:
Especially because you could get to this state by just requesting and then revoking in a normal legacy API flow? How do we revoke non-account permissions? So all the scopes would be left dangling without accounts?

Reply:
Like this case: https://github.com/MetaMask/metamask-extension/pull/26054/files#diff-118943e6acdedf444f19076e3615be646d9c19efb60eca649913f64d218dcf6bR219-R229

these optionalScopes (which could have been generated via the legacy `wallet_requestPermissions` method) are just left dangling without associated accounts in perpetuity